feat: add metrics module aligned with unification

plaguss opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR aligns the metric modules for a dataset with the unification strategies.
It includes a new `UnifiedAnnotationMetric` class to compute the `AnnotatorMetric` for a *unified* dataset.

Should be merged after #4175.

Closes #4263, #4034

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor (change restructuring the codebase without changing functionality)
- [ ] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [x] `tests/integration/client/feedbac/metrics/test_annotator_metrics.py`

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

plaguss committed:
chore: add metrics dependency

plaguss committed:
feat: add new metrics module

plaguss committed:
fix: remove unused data from tests and activate back parametrized tests

plaguss committed:
draft: initial work on metrics module

plaguss committed:
chore: merge with develop and fix conflicts

plaguss committed:
draft: initial work on metrics module

plaguss committed:
feat: return dict instead of defaultdict

plaguss committed:
feat: improve error message

plaguss committed:
feat: move prepare suggestions and reponses to a different function

plaguss committed:
test: tests for get_responses_per_user

plaguss committed:
chore: remove unused dependencies

plaguss committed:
feat: add check for metrics per question

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

plaguss committed:
feat: create utils module for common functions

plaguss committed:
feat: create agreement metrics module

plaguss committed:
feat: redirect imports to top level module

plaguss committed:
feat: move helper function to utils module

plaguss committed:
Merge branch 'feat/metrics' of github.com:argilla-io/argilla into feat/metrics

plaguss committed:
test: passing initial tests for metrics

plaguss committed:
feat: add docstrings for metrics module

plaguss committed:
Merge branch 'develop' of github.com:argilla-io/argilla into feat/metrics

plaguss committed:
refactor: include comments from code review

plaguss committed:
refactor: update distance functions to avoid casting to strings and computing edit_distance between them

plaguss committed:
fix: correct distance function for ranking questions

Commit:
Merge branch 'develop' of https://github.com/argilla-io/argilla into feat/metrics

Commit:
Merge branch 'feat/metrics' of https://github.com/argilla-io/argilla into feat/metrics

Commit:
doc: add examples in the docstrings

Commit:
doc: add reference docs for the metrics module

Commit:
doc: add initial docs for the metrics

Commit:
Merge branch 'develop' of https://github.com/argilla-io/argilla into feat/metrics

Commit:
chore: update changelog

Commit:
feat: add new function to gather unified responses and suggestions

Commit:
test: add tests for helper function

Commit:
test: add tests for unified annotator metrics

Commit:
feat: add redirection imports

Commit:
refactor: move common functionality to base class

Commit:
feat: add class to compute metrics for unified annotation

Commit:
wip: add method to compute the metrics from a feedback dataset

Commit:
refactor: update metric class name and correct circular imports

Commit:
feat: add repr method to metric classes

davidberenstein1957 commented on docs/_source/practical_guides/collect_responses.md:
````diff
@@ -120,6 +120,83 @@ Once you have unified your responses, you will have a dataset that's ready for [
+# [AgreementMetricResult(metric_name='alpha', result=0.0)]
+```
+
+Currently only [Krippendorf's alpha](https://en.wikipedia.org/wiki/Krippendorff%27s_alpha) is defined, for all the questions types except the `TextQuestion`. It applies to any number of annotators, with possible missing responses from them.
````
"with possible missing responses from them." is not very clear to me

davidberenstein1957 commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -627,3 +629,65 @@ def find_similar_records(
         )
         return []
+
+    def compute_annotator_metrics(
```
This looks good. perhaps we should also rename `unify_responses` to `compute_unified_responses`? I think aligning naming.

davidberenstein1957 reviewed (commented):
Hi @plaguss , this looks good already. If you want, feel free to also have a look at the structure of the code in general. I think it might be nice to structure the code in a similar way as the `metrics` module.

davidberenstein1957 committed:
Merge branch 'develop' into feat/metrics

Commit:
refactor: allow the uset to filter by response status

Commit:
refactor: remove unused kwargs for simplicity

Commit:
docs: add comments from code review

Commit:
Merge branch 'feat/metrics' of https://github.com/argilla-io/argilla into feat/metrics

Commit:
chore: merge and solve conflicts

Commit:
refactor: allow general filtering on the responses

Commit:
Merge branch 'feat/metrics' of https://github.com/argilla-io/argilla into refactor/align-unification-metrics

Commit:
feat: include sort_by and max_records to align with the codebase

Commit:
Merge branch 'feat/metrics' of https://github.com/argilla-io/argilla into refactor/align-unification-metrics

Commit:
feat: include metrics for multi label questions

Commit:
feat: include metric for ranking loss

davidberenstein1957 commented on docs/_source/practical_guides/collect_responses.md:
```diff
@@ -189,15 +189,16 @@ We plan on adding more support for other metrics so feel free to reach out on ou
 | spearman-r            |                |                    | ✔️              |                 |              |
 | gleu                  |                |                    |                |                 | ✔️            |
 | rouge                 |                |                    |                |                 | ✔️            |
+| ndcg-score            |                |                    |                | ✔️               | ✔️            |
```
I believe this one does not apply for `TextQuestion`

plaguss replied:
Ups, copy/paste error, will update it

Commit:
refactor: remove unused kwargs

Commit:
docs: remove wrong associated metric

Commit:
wip: ranking question with unification not working properly

Commit:
refactor: update unify_responses to compute_unified_responses to be on par with the compute_metrics

Commit:
fix: control for binary responses in label questions and missing suggestions

Commit:
feat: include filters to have on par agreement and annotator metrics

Commit:
docs: add reference to unification metrics

Commit:
docs: update reference for the new compute_unified_responses method name

Commit:
Merge branch 'develop' of https://github.com/argilla-io/argilla into refactor/align-unification-metrics

Commit:
docs: update metrics allowed

Commit:
feat: add the possibility to compute the annotator metrics using as reference both responses and suggestions

Commit:
Merge branch 'develop' of https://github.com/argilla-io/argilla into refactor/align-unification-metrics

Commit:
tests: add tests for the inversion of responses vs suggestions for annotator metrics

Commit:
feat: remove unnecesary check for submitted responses with the filters

Commit:
feat: included tests for remote feedback datasets and extra checks to ensure those work as expected

Commit:
feat: add specific classes for responses vs suggestions and viceversa

Commit:
fix: update typing for older python versions

Commit:
docs: add reference to the new specific classes

Commit:
feat: add metric computation from feedback datasetç

Commit:
fix: rename module

Commit:
fix: remove records without responses to avoid errors in agreement metrics

Commit:
fix: correct errors for users with different number of suggestions

Commit:
fix: remove extra ravel of nested structures that caused bad comparison

Commit:
fix: solved error in confusion matrix labels

Commit:
feat: include tqdm bar for to help with bigger datasets

davidberenstein1957 commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -418,15 +421,15 @@ def delete_metadata_properties(
         return deleted_metadata_properties if len(deleted_metadata_properties) > 1 else deleted_metadata_properties[0]
 
-    def unify_responses(
+    def compute_unified_responses(
```
can you add a depracation warning for this?

plaguss replied:
Sorry, I forgot it. It's done, I opened a new issue to test the deprecation and add notes in the documentation for that: #4322

plaguss replied:
Done

davidberenstein1957 commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -63,7 +64,9 @@
 
-class FeedbackDataset(ArgillaMixin, HuggingFaceDatasetMixin, FeedbackDatasetBase[FeedbackRecord], TaskTemplateMixin):
+class FeedbackDataset(
+    ArgillaMixin, HuggingFaceDatasetMixin, FeedbackDatasetBase[FeedbackRecord], TaskTemplateMixin, MetricsMixin
```
can you also create a unification mixin?

plaguss replied:
Done!

plaguss replied:
👍

Commit:
feat: deprecate curent unify_responses instead of removing it directly from the codebase

Commit:
feat: move unification methods to a mixin class

Commit:
Merge branch 'develop' into refactor/align-unification-metrics

Commit:
fix: add check for remote ranking questions

Commit:
fix: solve error with the ranking questions on unification

davidberenstein1957 commented on docs/_source/practical_guides/collect_responses.md:
```diff
@@ -120,6 +120,124 @@ Once you have unified your responses, you will have a dataset that's ready for [
+| ndcg-score            |                |                    |                | ✔️               |              |
+
+
+#### Unified annotation metrics
```
I think we should use `UnifiedAnnotatorMetric` to align with `AnnotatorMetric`

plaguss replied:
Done

davidberenstein1957 commented on docs/_source/reference/python/python_annotation_metrics.rst:
```diff
@@ -0,0 +1,59 @@
+.. autoclass:: argilla.client.feedback.metrics.annotator_metrics.AnnotatorMetric
+   :members: __init__, compute
+
+.. autoclass:: argilla.client.feedback.metrics.annotator_metrics.ResponsesMetric
```
maybe we can also unify the usage of plural vs singular here? In some cases we use plural and in others singular. 
`Annotator` vs `Suggestions`

davidberenstein1957 replied:
Intuitively, I would go for all singular.

plaguss replied:
Sure

davidberenstein1957 commented on docs/_source/practical_guides/collect_responses.md:
```diff
@@ -120,6 +120,124 @@ Once you have unified your responses, you will have a dataset that's ready for [
+| rouge                 |                |                    |                |                 | ✔️            |
+| ndcg-score            |                |                    |                | ✔️               |              |
+
+
```
Can you update the docs to include the Suggestion and ResponseMetric

plaguss replied:
On it!

plaguss replied:
Those will be the same for ResponsesMetric and SuggestionMetrics I guess.

Commit:
fix: align naming plural vs singular

Commit:
docs: updated docs using a real dataset example and the new class names

Commit:
fix: remove commented code

kursathalat committed:
fix: Fix typo in `SuggestionsMetric` (#4384)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

The module is imported as `SuggestionMetric` instead of
`SuggestionsMetric`, which causes an error. This PR fixes the typo.

Closes #4383 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

kursathalat committed:
docs: Improve info in `metrics` section (#4402)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This page can be improved with new info to demonstrate better the
metrics module. Also fixed typos.

Closes #4401 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

kursathalat committed:
feat: Align the docstrings with new `metrics` classification (#4409)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

As described in the issue, some parts of the docstrings should be
changed for better guidance. Also, if needed, class names can be
changed, though they look great at the moment.

Closes #4408 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing
functionality)
- [ ] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

kursathalat committed:
fix: Fix `prepare_dataset_for_annotation_task` to return correct format (#4397)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

As described in the issue, the method returns question name for all
records. In this PR, the question texts are returned so that the Alpha
is correctly corrected. The method can be further improved with a better
alternative.

Closes #4396 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

kursathalat committed:
docs: `end2end` tutorial on using `metrics` module (#4385)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Notes:
- Tutorial is not indexed properly yet.
- Some links to be corrected.
- Outputs can be changed for a better and readable version.

Closes #4186 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

davidberenstein1957 committed:
Merge branch 'develop' into refactor/align-unification-metrics

kursathalat committed:
feat: Add tests and docs for deprecated `unify_responses` method (#4412)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR adds tests and warning for deprecated `unify_responses` method.
All source code seems fine with no reference to the old method. Is there
any other part to change (especially in docs)?

Closes #4322 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing
functionality)
- [ ] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: plaguss <agustin@argilla.io>

plaguss committed:
feat: allow unified metrics to deal with the unification internally instead of the user

plaguss committed:
test: update test for the unified remote dataset

davidberenstein1957 committed:
chore: limit setfit version to <1.0.0

davidberenstein1957 committed:
Merge branch 'develop' into refactor/align-unification-metrics

davidberenstein1957 committed:
Merge branch 'develop' into refactor/align-unification-metrics

davidberenstein1957 committed:
refactor: renamed SuggestionMetric to ModelMetric
refactor: changed assumption that SuggestionMetric to default to responses as y_true
refactor: explicit varible usage for metrics to avoid inconsistencies
docs: updated docs to align with refac ModelMetric

davidberenstein1957 committed:
chore: update tests

davidberenstein1957 committed:
chore: fix faulty load from argilla

davidberenstein1957 committed:
docs: change naming

plaguss commented on docs/_source/practical_guides/collect_responses.md:
```diff
@@ -203,8 +203,8 @@ We obtain a `dict` where the keys contain the `user_id` of a given annotator and
-suggestions_metrics = dataset.compute_model_metrics(question_name="label", metric_names=["accuracy", "precision", "recall", "f1-score"])
-suggestions_metrics['00000000-0000-0000-0000-000000000001']
+mmodel_metrics = dataset.compute_model_metrics(question_name="label", metric_names=["accuracy", "precision", "recall", "f1-score"])
+mmodel_metrics['00000000-0000-0000-0000-000000000001']
```
typo in `mmodel_metrics`

plaguss reviewed (commented):
nice job!! 🙌

davidberenstein1957 committed:
chore: rename AnnotatorMetricResult to ModelMetricResult

davidberenstein1957 committed:
tests: fix failing tests

davidberenstein1957 committed:
chore: updated feedbackdataset records with metdata fixture
